File: nocternal-playz/android/theme_manager/src/main/kotlin/com/nocternal/playz/theme/ThemeSwitcher.kt
package com.nocternal.playz.theme

import com.nocternal.playz.model.AppSettings
import com.nocternal.playz.model.AudioSource
import com.nocternal.playz.model.GenreDefinition
import com.nocternal.playz.model.Mood
import com.nocternal.playz.model.ThemePreset
import com.nocternal.playz.model.Track
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** Why a theme switch was requested. Higher [priority] wins over lower until the context changes. */
enum class ThemeTrigger(val priority: Int) {
    TIME_OF_DAY(10),
    SOURCE_CHANGED(30),
    AI_MOOD_DETECTED(40),
    SONG_STARTED(60),
    GENRE_PLAYLIST_OPENED(80),
    GENRE_TILE_SELECTED(80),
    MANUAL(100),
}

sealed interface ThemeEvent {
    /** [queueGenreId] is set when the song plays from a genre playlist; that context wins over per-song detection. */
    data class SongStarted(val track: Track, val queueGenreId: String? = null) : ThemeEvent
    data class GenrePlaylistOpened(val genreId: String) : ThemeEvent
    data class MoodDetected(val mood: Mood, val confidence: Float, val hourOfDay: Int? = null) : ThemeEvent
    data class SourceChanged(val source: AudioSource, val nowPlaying: Track? = null) : ThemeEvent
    data class GenreTileSelected(val genreId: String) : ThemeEvent
    data class TimeTick(val hourOfDay: Int) : ThemeEvent
    data class ManualThemeSelected(val presetId: String) : ThemeEvent
    data object ManualLockReleased : ThemeEvent
}

data class ThemeDecision(
    val preset: ThemePreset,
    val genre: GenreDefinition?,
    val trigger: ThemeTrigger,
    val reason: String,
)

/**
 * Genre-based theme switching (spec §9):
 *
 *     val genre = detectGenre(currentSong)
 *     val theme = genreThemeMap[genre]
 *     applyTheme(theme)
 *
 * plus the rules that make it feel right in practice:
 *  - A manual theme choice locks auto switching until released.
 *  - A genre tile or genre playlist sets a context that per-song detection does not override.
 *  - Mood only switches the theme when the song's genre could not be detected.
 *  - Time-of-day only applies while nothing more specific is active.
 *  - Re-applying the same preset is a no-op (no flicker between songs of one genre).
 *  - A custom accent colour from settings always wins over the preset accent.
 */
class ThemeSwitcher(
    private val applier: ThemeApplier,
    private val settings: () -> AppSettings,
    private val detector: GenreDetector = GenreDetector(),
    private val genreThemeMap: Map<String, String> = GenreCatalog.genreThemeMap,
    private val moodConfidenceThreshold: Float = 0.6f,
) {
    private val _current = MutableStateFlow(
        ThemeDecision(ThemePresets.NOCTERNAL_DEFAULT, null, ThemeTrigger.TIME_OF_DAY, "default theme"),
    )
    val current: StateFlow<ThemeDecision> = _current.asStateFlow()

    private var manualLock = false
    private var contextGenreId: String? = null
    private var lastSongGenreKnown = false

    /** Handles an event; returns the decision if the theme changed, null if it was ignored or unchanged. */
    fun onEvent(event: ThemeEvent): ThemeDecision? {
        val s = settings()
        val decision: ThemeDecision = when (event) {
            is ThemeEvent.ManualThemeSelected -> {
                manualLock = true
                ThemeDecision(ThemePresets.byId(event.presetId), null, ThemeTrigger.MANUAL, "chosen by you")
            }
            ThemeEvent.ManualLockReleased -> {
                manualLock = false
                contextGenreId?.let { forGenre(it, ThemeTrigger.GENRE_PLAYLIST_OPENED, "back to playlist theme") }
                    ?: ThemeDecision(ThemePresets.NOCTERNAL_DEFAULT, null, ThemeTrigger.TIME_OF_DAY, "auto theme resumed")
            }
            else -> if (manualLock) null else autoDecision(event, s)
        } ?: return null

        if (decision.preset.id == _current.value.preset.id && event !is ThemeEvent.ManualThemeSelected) {
            _current.value = decision.copy(preset = _current.value.preset)
            return null
        }
        val preset = s.customAccent?.let { decision.preset.copy(accent = it) } ?: decision.preset
        applier.applyTheme(preset, applyEq = s.eqFollowsTheme)
        val applied = decision.copy(preset = preset)
        _current.value = applied
        return applied
    }

    fun detectGenre(track: Track): GenreDefinition? = detector.detect(track)?.genre

    private fun autoDecision(event: ThemeEvent, s: AppSettings): ThemeDecision? = when (event) {
        is ThemeEvent.GenreTileSelected -> {
            contextGenreId = event.genreId
            forGenre(event.genreId, ThemeTrigger.GENRE_TILE_SELECTED, "genre tile")
        }
        is ThemeEvent.GenrePlaylistOpened -> {
            contextGenreId = event.genreId
            forGenre(event.genreId, ThemeTrigger.GENRE_PLAYLIST_OPENED, "genre playlist")
        }
        is ThemeEvent.SongStarted -> onSongStarted(event, s)
        is ThemeEvent.MoodDetected -> {
            val weaker = _current.value.trigger.priority <= ThemeTrigger.AI_MOOD_DETECTED.priority
            if (s.autoThemeByMood && event.confidence >= moodConfidenceThreshold && (weaker || !lastSongGenreKnown) && contextGenreId == null) {
                val genre = detector.forMood(event.mood, event.hourOfDay)
                forGenre(genre.id, ThemeTrigger.AI_MOOD_DETECTED, "mood looks ${event.mood.label.lowercase()}")
            } else null
        }
        is ThemeEvent.SourceChanged -> {
            contextGenreId = null
            val genre = event.nowPlaying?.let { detector.detect(it)?.genre }
            if (genre != null) forGenre(genre.id, ThemeTrigger.SOURCE_CHANGED, "${event.source.label} · ${genre.displayName}")
            else ThemeDecision(sourcePreset(event.source), null, ThemeTrigger.SOURCE_CHANGED, "switched to ${event.source.label}")
        }
        is ThemeEvent.TimeTick -> {
            if (s.autoThemeByTime && _current.value.trigger.priority <= ThemeTrigger.TIME_OF_DAY.priority) {
                val genre = TimeOfDayThemes.genreForHour(event.hourOfDay)
                forGenre(genre.id, ThemeTrigger.TIME_OF_DAY, "time of day")
            } else null
        }
        is ThemeEvent.ManualThemeSelected, ThemeEvent.ManualLockReleased -> null
    }

    private fun onSongStarted(event: ThemeEvent.SongStarted, s: AppSettings): ThemeDecision? {
        event.queueGenreId?.let { contextGenreId = it }
        if (event.queueGenreId == null && contextGenreId != null) contextGenreId = null
        contextGenreId?.let { id ->
            lastSongGenreKnown = true
            return forGenre(id, ThemeTrigger.GENRE_PLAYLIST_OPENED, "playing from genre playlist")
        }
        if (!s.autoThemeByGenre) return null
        val match = detector.detect(event.track)
        lastSongGenreKnown = match != null
        return match?.let { forGenre(it.genre.id, ThemeTrigger.SONG_STARTED, it.reason) }
    }

    private fun forGenre(genreId: String, trigger: ThemeTrigger, reason: String): ThemeDecision? {
        val genre = GenreCatalog.byId(genreId) ?: return null
        val presetId = genreThemeMap[genreId] ?: genre.themePresetId
        return ThemeDecision(ThemePresets.byId(presetId), genre, trigger, reason)
    }

    private fun sourcePreset(source: AudioSource): ThemePreset = when (source) {
        AudioSource.LOCAL -> ThemePresets.NOCTERNAL_DEFAULT
        AudioSource.YOUTUBE -> ThemePresets.CYBER_RED_PULSE
        AudioSource.RADIO -> ThemePresets.BLUE_PURPLE_GALAXY
    }
}

File: nocternal-playz/ios/NocternalKit/Sources/ThemeManager/ThemeSwitcher.swift
import Foundation
import NocternalModel

// The four sinks from spec §9 ("ThemeManager.setAccentColor … BackgroundManager.setStyle").
public protocol ThemeManaging: AnyObject {
    func setAccentColor(_ c: NeonColor)
    func setSecondaryAccentColor(_ c: NeonColor)
    func setGlowIntensity(_ v: Double)
    func setThemeMode(_ m: ThemeMode)
}
public protocol LightingEngine: AnyObject {
    func setEdgeMode(_ m: EdgeLightingMode)
    func setEdgeStyle(thickness: Double, brightness: Double)
    func setLightBarAnimation(_ a: LightingAnimation)
}
public protocol EQManaging: AnyObject { func applyPreset(_ p: EqPreset) }
public protocol BackgroundManaging: AnyObject { func setStyle(_ s: BackgroundStyle) }

public final class ThemeApplier {
    let theme: ThemeManaging, lighting: LightingEngine, eq: EQManaging, background: BackgroundManaging
    public init(theme: ThemeManaging, lighting: LightingEngine, eq: EQManaging, background: BackgroundManaging) {
        self.theme = theme; self.lighting = lighting; self.eq = eq; self.background = background
    }
    public func applyTheme(_ t: ThemePreset, applyEq: Bool = true) {
        theme.setAccentColor(t.accent)
        theme.setSecondaryAccentColor(t.secondaryAccent)
        theme.setGlowIntensity(t.glowIntensity)
        lighting.setEdgeMode(t.edgeLightingMode)
        lighting.setEdgeStyle(thickness: t.edgeThickness, brightness: t.edgeBrightness)
        lighting.setLightBarAnimation(t.lightBarAnimation)
        if applyEq { eq.applyPreset(EqPresets.byId(t.eqPresetId)) }
        background.setStyle(t.backgroundStyle)
    }
}

public enum ThemeTrigger: Int, Sendable {
    case timeOfDay = 10, sourceChanged = 30, aiMoodDetected = 40, songStarted = 60, genrePlaylistOpened = 80, genreTileSelected = 81, manual = 100
    var priority: Int { self == .genreTileSelected ? 80 : rawValue }
}

public enum ThemeEvent: Sendable {
    case songStarted(Track, queueGenreId: String?)
    case genrePlaylistOpened(String)
    case moodDetected(Mood, confidence: Double, hour: Int?)
    case sourceChanged(AudioSource, nowPlaying: Track?)
    case genreTileSelected(String)
    case timeTick(hour: Int)
    case manualThemeSelected(String)
    case manualLockReleased
}

public struct ThemeDecision: Sendable {
    public let preset: ThemePreset
    public let genre: GenreDefinition?
    public let trigger: ThemeTrigger
    public let reason: String
}

/// Genre-based theme switching (spec §9) with the same rules as the Kotlin `ThemeSwitcher`.
public final class ThemeSwitcher {
    private let applier: ThemeApplier
    private let settings: () -> AppSettings
    private let detector: GenreDetector
    private let moodThreshold = 0.6
    public private(set) var current = ThemeDecision(preset: ThemePresets.nocternalDefault, genre: nil, trigger: .timeOfDay, reason: "default theme")
    public var onChange: ((ThemeDecision) -> Void)?
    private var manualLock = false
    private var contextGenreId: String?
    private var lastSongGenreKnown = false

    public init(applier: ThemeApplier, settings: @escaping () -> AppSettings, detector: GenreDetector = GenreDetector()) {
        self.applier = applier; self.settings = settings; self.detector = detector
    }

    @discardableResult
    public func handle(_ event: ThemeEvent) -> ThemeDecision? {
        let s = settings()
        let decision: ThemeDecision?
        switch event {
        case .manualThemeSelected(let id):
            manualLock = true
            decision = ThemeDecision(preset: ThemePresets.byId(id), genre: nil, trigger: .manual, reason: "chosen by you")
        case .manualLockReleased:
            manualLock = false
            decision = contextGenreId.flatMap { forGenre($0, .genrePlaylistOpened, "back to playlist theme") }
                ?? ThemeDecision(preset: ThemePresets.nocternalDefault, genre: nil, trigger: .timeOfDay, reason: "auto theme resumed")
        default:
            decision = manualLock ? nil : auto(event, s)
        }
        guard let d = decision else { return nil }
        var isManual = false
        if case .manualThemeSelected = event { isManual = true }
        if d.preset.id == current.preset.id && !isManual {
            current = ThemeDecision(preset: current.preset, genre: d.genre, trigger: d.trigger, reason: d.reason)
            return nil
        }
        var preset = d.preset
        if let accent = s.customAccent { preset.accent = accent }
        applier.applyTheme(preset, applyEq: s.eqFollowsTheme)
        current = ThemeDecision(preset: preset, genre: d.genre, trigger: d.trigger, reason: d.reason)
        onChange?(current)
        return current
    }

    private func auto(_ event: ThemeEvent, _ s: AppSettings) -> ThemeDecision? {
        switch event {
        case .genreTileSelected(let id):
            contextGenreId = id
            return forGenre(id, .genreTileSelected, "genre tile")
        case .genrePlaylistOpened(let id):
            contextGenreId = id
            return forGenre(id, .genrePlaylistOpened, "genre playlist")
        case .songStarted(let track, let queueGenreId):
            if let q = queueGenreId { contextGenreId = q } else { contextGenreId = nil }
            if let id = contextGenreId { lastSongGenreKnown = true; return forGenre(id, .genrePlaylistOpened, "playing from genre playlist") }
            guard s.autoThemeByGenre else { return nil }
            let m = detector.detect(track)
            lastSongGenreKnown = m != nil
            return m.flatMap { forGenre($0.genre.id, .songStarted, $0.reason) }
        case .moodDetected(let mood, let confidence, let hour):
            let weaker = current.trigger.priority <= ThemeTrigger.aiMoodDetected.priority
            guard s.autoThemeByMood, confidence >= moodThreshold, weaker || !lastSongGenreKnown, contextGenreId == nil else { return nil }
            return forGenre(detector.forMood(mood, hour: hour).id, .aiMoodDetected, "mood looks \(mood.label.lowercased())")
        case .sourceChanged(let source, let nowPlaying):
            contextGenreId = nil
            if let g = nowPlaying.flatMap({ detector.detect($0)?.genre }) { return forGenre(g.id, .sourceChanged, "\(source.label) · \(g.displayName)") }
            let preset: ThemePreset = source == .youtube ? ThemePresets.cyberRedPulse : source == .radio ? ThemePresets.bluePurpleGalaxy : ThemePresets.nocternalDefault
            return ThemeDecision(preset: preset, genre: nil, trigger: .sourceChanged, reason: "switched to \(source.label)")
        case .timeTick(let hour):
            guard s.autoThemeByTime, current.trigger.priority <= ThemeTrigger.timeOfDay.priority else { return nil }
            return forGenre(TimeOfDayThemes.genreForHour(hour).id, .timeOfDay, "time of day")
        case .manualThemeSelected, .manualLockReleased:
            return nil
        }
    }

    private func forGenre(_ id: String, _ trigger: ThemeTrigger, _ reason: String) -> ThemeDecision? {
        guard let g = GenreCatalog.byId(id) else { return nil }
        return ThemeDecision(preset: ThemePresets.byId(GenreCatalog.genreThemeMap[id] ?? g.themePresetId), genre: g, trigger: trigger, reason: reason)
    }
}
